### ISSUE TITLE
How to access specific weight value  in addition to loss for calculating weight update

### ISSUE THREAD

**Issue by mustgoplay** [fchollet/keras#4551]
I want weight changes in one of my layers to be dependent on the current weight and the outputs.  For example, let's say we're looking at xor and the input is 0 1 and the output is 1.   Let's say there are no hidden layers and no bias and no sigmoid/relu/etc.  This is unrealistic and very simple but is being used to explain the problem.  Let's say the weights are w00=0.25 and w10=0.75 but let's only look at the 0.75 weight.  I want the update to that weight to not only be dependent on the error (1-0.75) but also on w10.  How do I access the 0.75 value specifically during the training with respect to the loss?  

The nearest I can figure is that I would need to use an optimizer and that the value is in the params variable but I can't tell what's happening when a whole batch is being presented (i.e. how do isolate the single input) and I can't tell how to access the param to use in the update calculation.

Please make sure that the boxes below are checked before you submit your issue. Thank you!

- [x ] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [ x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [ x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by patyork** [comment]
This would probably fall under the umbrella of a (weight) [regularizer](https://keras.io/regularizers/). Depending on exactly how you wish the current weight value to affect the updated weight value, you may already be looking at an L1 or L2 weight regularizer, which Keras has built in.

If you are looking to implement a different regularizer other than L1/L2, you [should reference the current weight regularizer](https://github.com/fchollet/keras/blob/master/keras/regularizers.py#L66) types to create your own. You'll need a bit of an understanding of using either Theano or TensorFlow, and how the operations apply across matrices/weights, as there is rarely a time when you will be iterating over each weight and applying a function so there is little need to isolate any single weight values.

**Comment by mustgoplay** [comment]
Thank you.  

I'm actually trying to do my own weight updates so I need to know the actual weight or be able to access it while accessing the loss in order to determine the weight update.    I.e. not just regularize the output.

Looking at the regularizers, are the weights then held in self.p?   It's looking like they are only changing the output of the layer and not actually updating the weights.  Is that correct?

If so any other suggestions please?

Thanks

**Comment by patyork** [comment]
In the context of the weight regularizer, you can access them via the self.p attribute.

However, if you are going to create your own weight update/training process, you can utilize the `get_weights()` and `set_weights(weights)` methods on each layer. See at the top [here](https://keras.io/layers/about-keras-layers/).

**Comment by mustgoplay** [comment]
Hmm.  I didn't think you could access those during training because all the examples I saw had them being used outside the model for loading in previously saved weights or just printing the weights.  I'll play around with this and see if it'll work.

After your suggestions I'm trying to figure out how all the sub-layer functions (objective, regularizer, optimizer) interact including.   I can't find a good tutorial for dummies on how to write a complex custom layer that utilizes these things.

Thank you.

**Comment by mustgoplay** [comment]
OK, trying to write my own optimizer to start, then can work on the regularizer and loss functions.  I tried a bunch of variations (see commented out prints) and I can't seem to get get_values or get_wieghts to work.  I just simply copied the SGD optimizer, renamed it TestOptimizer and added the print statements.  Any ideas please on how to see the values of the tensors please?

`class TestOptimizer(Optimizer):
    '''Stochastic gradient descent, with support for momentum,
    learning rate decay, and Nesterov momentum.

    # Arguments
        lr: float >= 0. Learning rate.
        momentum: float >= 0. Parameter updates momentum.
        decay: float >= 0. Learning rate decay over each update.
        nesterov: boolean. Whether to apply Nesterov momentum.
    '''

    def __init__(self, lr=0.01, momentum=0., decay=0.,
                 nesterov=False, **kwargs):
      super(TestOptimizer, self).__init__(**kwargs)
      self.__dict__.update(locals())
      self.iterations = K.variable(0.)
      self.lr = K.variable(lr)
      self.momentum = K.variable(momentum)
      self.decay = K.variable(decay)

    def get_updates(self, params, constraints, loss):
      grads = self.get_gradients(loss, params)
      lr = self.lr * (1. / (1. + self.decay * self.iterations))
      self.updates = [K.update_add(self.iterations, 1)]

      # momentum
      shapes = [x.shape for x in K.batch_get_value(params)]
      moments = [K.zeros(shape) for shape in shapes]
      self.weights = [self.iterations] + moments
      for p, g, m in zip(params, grads, moments):
        v = self.momentum * m - lr * g  # velocity
        self.updates.append(K.update(m, v))

        if self.nesterov:
          new_p = p + self.momentum * v - lr * g
        else:
          new_p = p + v

        # apply constraints
        if p in constraints:
          c = constraints[p]
          new_p = c(new_p)

        self.updates.append(K.update(p, new_p))

        #self.params = params
        #self.loss = loss
        #print('params', K.get_value(self.params))
        #print('loss', K.get_value(self.loss))
        #print('updates', K.get_value(self.updates))
        #print('weights', K.get_value(self.weights))
        print('weights:', self.get_weights())
      return self.updates

    def get_config(self):
      config = {'lr': float(K.get_value(self.lr)),
                'momentum': float(K.get_value(self.momentum)),
                'decay': float(K.get_value(self.decay)),
                'nesterov': self.nesterov}
      base_config = super(SGD, self).get_config()
      return dict(list(base_config.items()) + list(config.items()))
`
Thanks

**Comment by mustgoplay** [comment]
This one kind of worked:

        print('weights', K.batch_get_value(params))

But it shows the exact same weights twice regardless of how many epochs there are so it's only showing the initial weights and not actually showing the changes to the wieghts in real-time.
